File: solutions/2026-08-22/cpp/2026-08-22-check-divisibility-by-digit-sum-and-product.cpp
// Problem: Check Divisibility by Digit Sum and Product
// Link: https://leetcode.com/problems/check-divisibility-by-digit-sum-and-product/
//
// Approach:
// To solve this problem, we need to calculate the sum and product of the digits of the given integer `n`.
// We can iterate through the digits of `n` by repeatedly taking the number modulo 10 to get the last digit,
// adding it to the sum, and multiplying it with the product. We then divide `n` by 10 to remove the last digit.
// This process continues until `n` becomes 0.
// After calculating the digit sum and digit product, we add them together to get the total divisor.
// Finally, we check if the original number `n` is divisible by this total divisor.
//
// Time Complexity:
// The time complexity is O(log10(n)), where n is the input integer. This is because the number of digits in n
// is proportional to log10(n), and we iterate through each digit once.
//
// Space Complexity:
// The space complexity is O(1) because we are using a fixed amount of extra space for variables regardless of the input size.

#include <iostream>

class Solution {
public:
    bool checkDivisibility(int n) {
        // Store the original value of n to perform the divisibility check at the end.
        int original_n = n;

        // Initialize variables to store the sum and product of digits.
        int digit_sum = 0;
        int digit_product = 1; // Initialize product to 1, as multiplying by 0 would make the product 0.

        // Handle the case where n is 0. Although constraints say n >= 1, it's good practice.
        if (n == 0) {
            // A number is divisible by 0 only if it's 0.
            // The sum of digits for 0 is 0, product is 0. Total divisor is 0.
            return true; // 0 is divisible by 0.
        }

        // Iterate through the digits of n.
        while (n > 0) {
            // Get the last digit of n.
            int digit = n % 10;

            // Add the digit to the sum.
            digit_sum += digit;

            // Multiply the digit with the product.
            digit_product *= digit;

            // Remove the last digit from n by integer division.
            n /= 10;
        }

        // Calculate the total divisor by summing the digit sum and digit product.
        int total_divisor = digit_sum + digit_product;

        // Check if the original number is divisible by the total divisor.
        // We also need to ensure that the total_divisor is not zero to avoid division by zero.
        // If total_divisor is zero, it means either digit_sum and digit_product are both zero,
        // which only happens if n was initially 0 (already handled).
        // For positive n, digit_sum will always be >= 1, so total_divisor will be >= 1.
        if (total_divisor == 0) {
            // This case should not be reachable given the constraint n >= 1.
            // If it were reachable, and original_n was also 0, it would be true.
            // If original_n was not 0, it would be false.
            return original_n == 0;
        }

        // Return true if original_n is perfectly divisible by total_divisor, false otherwise.
        return original_n % total_divisor == 0;
    }
};

// int main() {
//     Solution sol;
//     std::cout << "Input: 99, Output: " << std::boolalpha << sol.checkDivisibility(99) << std::endl; // Expected: true
//     std::cout << "Input: 23, Output: " << std::boolalpha << sol.checkDivisibility(23) << std::endl; // Expected: false
//     std::cout << "Input: 10, Output: " << std::boolalpha << sol.checkDivisibility(10) << std::endl; // Sum=1, Prod=0, Total=1. 10%1 == 0. Expected: true
//     std::cout << "Input: 1, Output: " << std::boolalpha << sol.checkDivisibility(1) << std::endl;   // Sum=1, Prod=1, Total=2. 1%2 != 0. Expected: false
//     std::cout << "Input: 12, Output: " << std::boolalpha << sol.checkDivisibility(12) << std::endl; // Sum=3, Prod=2, Total=5. 12%5 != 0. Expected: false
//     std::cout << "Input: 100, Output: " << std::boolalpha << sol.checkDivisibility(100) << std::endl; // Sum=1, Prod=0, Total=1. 100%1 == 0. Expected: true
//     return 0;
// }
```

File: solutions/2026-08-22/java/2026-08-22-check-divisibility-by-digit-sum-and-product.java
```java
// Problem Summary:
// Check if a given positive integer 'n' is divisible by the sum of its digit sum and digit product.
// Link: https://leetcode.com/problems/check-divisibility-by-digit-sum-and-product/
//
// Approach:
// 1. Iterate through the digits of the number 'n'.
// 2. For each digit, add it to a running sum (digit_sum) and multiply it into a running product (digit_product).
// 3. Handle the case where a digit is 0 to avoid multiplying the product by 0 and making it always 0. If a 0 is encountered, the digit product becomes 0.
// 4. Calculate the total divisor by adding digit_sum and digit_product.
// 5. Check if 'n' is divisible by the total_divisor.
//
// Time Complexity Analysis:
// The time complexity is O(log10(n)) because we iterate through the digits of 'n'. The number of digits in 'n' is proportional to log10(n).
//
// Space Complexity Analysis:
// The space complexity is O(1) because we only use a few constant variables to store the digit sum, digit product, and temporary digits.
class Solution {
    public boolean checkDivisibility(int n) {
        // Store the original number to perform the divisibility check later.
        int original_n = n;

        // Initialize variables to store the sum and product of digits.
        int digit_sum = 0;
        int digit_product = 1;
        boolean has_zero = false; // Flag to check if the number contains a zero digit.

        // Loop through each digit of the number.
        while (n > 0) {
            // Get the last digit.
            int digit = n % 10;

            // Add the digit to the sum.
            digit_sum += digit;

            // Multiply the digit into the product, but only if it's not zero.
            // If a zero digit is present, the product will effectively become zero for divisibility purposes later,
            // as per the problem's implicit logic or to avoid division by zero if the product becomes zero.
            if (digit != 0) {
                digit_product *= digit;
            } else {
                has_zero = true;
            }

            // Remove the last digit from the number.
            n /= 10;
        }

        // If the original number contained a zero digit, the effective product for the sum is 0.
        // However, the problem implies we should still consider the product of non-zero digits.
        // If the number itself is 0, digit_sum would be 0 and digit_product would be 1. But constraints say n >= 1.
        // If digit_product is 1 and has_zero is true, it means the number was something like 10 or 100 etc.
        // In such cases, the product of digits is 0.
        if (has_zero) {
            digit_product = 0;
        }

        // Calculate the total divisor by summing the digit sum and digit product.
        int total_divisor = digit_sum + digit_product;

        // If the total divisor is 0, it means digit_sum was 0 and digit_product was 0.
        // This can only happen if n itself was 0, but the constraints state n >= 1.
        // So, we don't need to explicitly check for division by zero if n >= 1.
        // If n was 10, digit_sum=1, digit_product=0, total_divisor=1. 10 % 1 == 0.
        // If n was 1, digit_sum=1, digit_product=1, total_divisor=2. 1 % 2 != 0.

        // Check if the original number is divisible by the total divisor.
        return original_n % total_divisor == 0;
    }
}
```